src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs

```diff
@@ -162,109 +189,205 @@ impl HummockVersionExt for HummockVersion {
+                "new max commit epoch {} is older than the current max commit epoch {}",
+                version_delta.max_committed_epoch,
+                self.max_committed_epoch
             );
+            if self.max_committed_epoch < version_delta.max_committed_epoch {
+                // `max_committed_epoch` increases. It must be a `commit_epoch`
+                let LevelDeltasSummary {
+                    delete_sst_levels,
+                    delete_sst_ids_set,
+                    insert_sst_level_id,
+                    insert_sub_level_id,
+                    insert_table_infos,
+                } = summary;
+                assert_eq!(
+                    0, insert_sst_level_id,
+                    "we should only add to L0 when we commit an epoch. Inserting into {} {:?}",
+                    insert_sst_level_id, insert_table_infos
+                );
+                assert!(
+                    delete_sst_levels.is_empty() && delete_sst_ids_set.is_empty(),
+                    "no sst should be deleted when committing an epoch"
+                );
+                add_new_sub_level(
+                    levels.l0.as_mut().unwrap(),
+                    insert_sub_level_id,
+                    insert_table_infos,
+                );
+            } else {
+                // `max_committed_epoch` is not changed. The delta is caused by compaction.
+                levels.apply_compact_ssts(summary, false);
+            }
         }
         self.id = version_delta.id;
         self.max_committed_epoch = version_delta.max_committed_epoch;
         self.max_current_epoch = version_delta.max_current_epoch;
         self.safe_epoch = version_delta.safe_epoch;
     }
+}
 
-    fn apply_compact_ssts(
-        &mut self,
-        compaction_group_id: CompactionGroupId,
-        delete_sst_levels: &[u32],
-        delete_sst_ids_set: &HashSet<u64>,
-        insert_sst_level: u32,
-        insert_sub_level_id: u64,
-        insert_table_infos: Vec<SstableInfo>,
-    ) {
-        if let Some(levels) = self.levels.get_mut(&compaction_group_id) {
-            for level_idx in delete_sst_levels {
-                if *level_idx == 0 {
-                    for level in &mut levels.l0.as_mut().unwrap().sub_levels {
-                        level_delete_ssts(level, delete_sst_ids_set);
-                    }
-                } else {
-                    let idx = *level_idx as usize - 1;
-                    level_delete_ssts(&mut levels.levels[idx], delete_sst_ids_set);
+pub trait HummockLevelsExt {
+    fn get_level0(&self) -> &OverlappingLevel;
+    fn get_level(&self, idx: usize) -> &Level;
+    fn get_level_mut(&mut self, idx: usize) -> &mut Level;
+    fn apply_compact_ssts(&mut self, summary: LevelDeltasSummary, local_related_only: bool);
+}
+
+impl HummockLevelsExt for Levels {
+    fn get_level0(&self) -> &OverlappingLevel {
+        self.l0.as_ref().unwrap()
+    }
+
+    fn get_level(&self, level_idx: usize) -> &Level {
```

Comment:
guess we'd better not minus here avoid confusion

Reply:
The code is simply moved from L305 in the current main branch in the same file. We'd better keep it consistent with previous code.